Issue #7749: Update AbstractChecks to log DetailAST - MissingSwitchDe…

Abhishek-kumar09 opened the pull request:
Fixes Issue #7749 

[Diff report](https://abhishek-kumar09.github.io/7749/reports/diff/index.html)

rdiachenko commented:
@Abhishek-kumar09 please update your first post with a link to diff report. This is an item 4 from https://github.com/checkstyle/checkstyle/issues/5777, please recheck other items as well.

Abhishek-kumar09 commented:
> @Abhishek-kumar09 please update your first post with a link to diff report. This is an item 4 from #5777, please recheck other items as well.

Ok, I am doing so.

Abhishek-kumar09 commented:
> @Abhishek-kumar09 please update your first post with a link to diff report. This is an item 4 from #5777, please recheck other items as well.

@rdiachenko Its done :+1: , quite a long process with dangling internet connection.

rdiachenko commented:
@Abhishek-kumar09 , could you please also provide my_check.xml you used for report generation. It is very weird that none of the listed projects have violation for the check.

Changes for the code itself look fine for me. @pbludov could you recheck please?

strkkk commented:
@Abhishek-kumar09 it looks like you forgot to include your check in diff report, now report is invalid. 
There should be even numbers of changes for each project (if it is odd, there are some new violations and they should be analyzed)
Please also rebase on latest master to resolve wercker issue.

Abhishek-kumar09 commented:
Ok , Thanks I will do so.

Abhishek-kumar09 commented:
> @Abhishek-kumar09 it looks like you forgot to include your check in diff report, now report is invalid.
> There should be even numbers of changes for each project (if it is odd, there are some new violations and they should be analyzed)
> Please also rebase on latest master to resolve wercker issue.

@strkkk  I have did it. Updated the diff report and rebased the branch with latest master.

strkkk commented on src/it/resources/org/checkstyle/suppressionxpathfilter/missingswitchdefault/SuppressionXpathRegressionMissingSwitchDefaultOne.java:
```diff
@@ -0,0 +1,14 @@
+public class SuppressionXpathRegressionMissingSwitchDefaultOne {
+    public static void main(String[] args) {
+        int key = 2;
+        // BadExample
```
remove

strkkk commented on src/it/resources/org/checkstyle/suppressionxpathfilter/missingswitchdefault/SuppressionXpathRegressionMissingSwitchDefaultOne.java:
```diff
@@ -0,0 +1,14 @@
+package org.checkstyle.suppressionxpathfilter.missingswitchdefault;
+
+public class SuppressionXpathRegressionMissingSwitchDefaultOne {
+    public static void main(String[] args) {
```
why main? please change to some other method

strkkk commented on src/it/resources/org/checkstyle/suppressionxpathfilter/missingswitchdefault/SuppressionXpathRegressionMissingSwitchDefaultTwo.java:
```diff
@@ -0,0 +1,18 @@
+package org.checkstyle.suppressionxpathfilter.missingswitchdefault;
+
+public class SuppressionXpathRegressionMissingSwitchDefaultTwo {
+    public static void main(String[] args) {
```
here too

Abhishek-kumar09 replied:
ok

strkkk commented on src/it/resources/org/checkstyle/suppressionxpathfilter/missingswitchdefault/SuppressionXpathRegressionMissingSwitchDefaultTwo.java:
```diff
@@ -0,0 +1,18 @@
+package org.checkstyle.suppressionxpathfilter.missingswitchdefault;
+
+public class SuppressionXpathRegressionMissingSwitchDefaultTwo {
```
Please also update unit tests with example of nested switch

Abhishek-kumar09 replied:
Should I nest another switch inside another Switch's case?

Abhishek-kumar09 replied:
I had nested a switch inside a switch default.

strkkk replied:
It can be both.
In report I saw examples of nested switch, so it is good to have them in tests

strkkk replied:
@Abhishek-kumar09 I was saying about unit tests, not xpath regression tests. Please correct

rnveach commented:
> There should be even numbers of changes for each project (if it is odd, there are some new violations and they should be analyzed)

While this is true if there is only one case, please note that if report has multiple single ones, it could turn the count even and may go unnoticed. 

A sure way to ensure there are no new violations is to examine `Number of unique base messages reported below` and `Number of unique patch messages reported below` for each project. They should match, otherwise there is a new/removed violation.

strkkk commented:
> While this is true if there is only one case, please note that if report has multiple single ones, it could turn the count even and may go unnoticed.

Good point, thank you.

>A sure way to ensure there are no new violations is to examine Number of unique base messages reported below and Number of unique patch messages reported below for each project. They should match, otherwise there is a new/removed violation.

May be it makes sense to post this information on main report page, e.g. 
```
android-launcher (total: 42, new: 21, removed 21)
```
it will be easier to detect where the issue can be instead of clicking on each of 10+ project links.

rnveach commented:
> May be it makes sense to post this information on main report page

It does make sense. Please make an issue for it.

Abhishek-kumar09 commented:
Its done, Diff report is also updated.

Abhishek-kumar09 commented:
OK, I have updated the unit tests, Third XpathRegression test is still there in the latest commit, Should I remove it?

strkkk commented:
> Third XpathRegression test is still there in the latest commit, Should I remove it?

@Abhishek-kumar09 I think yes, xpath regression tests should not be copy of unit tests.
Please also rebase to latest master and resolve merge conflicts

Abhishek-kumar09 commented:
> > Third XpathRegression test is still there in the latest commit, Should I remove it?
> 
> @Abhishek-kumar09 I think yes, xpath regression tests should not be copy of unit tests.
> Please also rebase to latest master and resolve merge conflicts

OK, I have removed it. Rebased the branch with current master

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -19,4 +19,27 @@ public void foo() {
+        switch (i) {
+            case 1: i++; break;
+            case 2: i--; break;
+            default: switch (i++) {
```
please move nested switch to next line

Abhishek-kumar09 replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -19,4 +19,27 @@ public void foo() {
         }
     }
+
+    public void nestedSwitch1() {
```
please add `// violation` comments on lines with violation

Abhishek-kumar09 replied:
Done, I have added comments only in unitTest inputs.

Abhishek-kumar09 replied:
Updated the Diff report.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -19,4 +19,29 @@ public void foo() {
+            case 1: i++; break;
+            case 2: i--; break;
+            default:
+                switch (i++) { // violation, switch without "default" clause
```
there is not need for detailed comment (it is not documentation), just `// violation` is enough.
This comment is needed to make it easier to detect violations.
Please correct.

Abhishek-kumar09 replied:
OK

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -19,4 +19,29 @@ public void foo() {
+        int i = 1, j = 2;
+        switch(i) {
+            case 1:
+                switch (j) { // violation, switch without "default" clause
```
here too

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -19,4 +19,29 @@ public void foo() {
         case 2: i--; break;
```
this switch also produces violation, please add comment here too as well.

Abhishek-kumar09 replied:
OK

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingswitchdefault/InputMissingSwitchDefault.java:
```diff
@@ -4,19 +4,44 @@
-        case 1: i++; break;
-        case 2: i--; break;
-        default: return;
+            case 1: i++; break;
```
Why did test cases in inputs change?

Abhishek-kumar09 replied:
I added tab spaces to give a consistent and even look to whole file, the already added inputTestCases were not changed by me.

romani replied:
please move any unrelated to issue fixes (formatting, indentations to separate commit of the same PR)

Abhishek-kumar09 commented:
> looks good, but formatting changes should be moved to separate commit(first commit before commit with functional changes) with message `minor: fix indentation in test inputs`.

Ok done, split the code changes into two commits 
one for minor fixes
other for issue resolution.
